openedx/openedx-platform
# [FC-0009] Copy/Paste associated static assets

## Description

This implements https://github.com/openedx/modular-learning/issues/16 so that users can copy and paste components between different courses, with static assets coming along "automatically" as needed.

## Supporting information

See https://github.com/openedx/modular-learning/issues/16

This also includes a small fix for https://github.com/openedx/modular-learning/issues/66

## Testing instructions

Prequisite: enable the new copy-paste feature:
1. Go to http://studio.local.overhang.io:8001/admin/waffle/flag/ (or devstack/sandbox equivalent)
2. Add a new flag, `contentstore.enable_copy_paste_feature` and enable it for Everyone

Instructions:
1. You need two courses.
2. In course A, create some components that reference static assets using the `/static/blah.ext` format. The files should be added to course A's "Files & Uploads". Ideally:
    * A Text/HTML component that embeds an image or links to a PDF.
    * A drag and drop component with a custom image.
    * A video with a custom transcript file uploaded via its settings.
    * A Problem component that depends on `python_lib.zip` but doesn't directly reference it ([example](https://github.com/mitodl/mitx-grading-library) from MIT; though note that the sample `course` contains two symlinks which must be replaced with hard links before it can be imported, if you want to import their sample course).
3. Copy these components and paste them into Course B.
4. Verify that the static assets now appear in Course B's "Files & Uploads" and that the components work as before.
5. Also test what happens if Course B already has an existing file with the same name but different contents (a warning should be shown).

6. This also includes a fix for https://github.com/openedx/modular-learning/issues/66 - please set an HTML block to "raw" editor then copy and paste it, and ensure the raw editor is saved.

~~Known issue: in the destination course's "Files & Uploads" page, no thumbnail images will be shown for automatically-copied assets. I'm not sure why but can investigate that in a future PR.~~ Working now!

## Deadline

None

## TODOs

- [x] Implement copying of static assets from "Files & Uploads"
- [x] Implement copying of static assets from the per-XBlock filesystem (e.g. video transcripts)
- [x] Implement pluggable filter to control what gets copied
- [x] Implement filter to avoid copying very large assets
- [x] Save static assets into the new course on paste
- [x] Display a notification to the user about new files or conflicting files
- [x] ~~Display the list of assets in the paste preview~~ (excluded from UX design for now)
- [x] Implement copying of `python_lib.zip` or at least store its hash
- [x] Add more test cases

Private-ref: MNG-3723
